packages/e2e-tests/specs/keyboard-navigable-blocks.test.js

```diff
@@ -0,0 +1,165 @@
+/**
+ * WordPress dependencies
+ */
+import {
+	createNewPost,
+	insertBlock,
+} from '@wordpress/e2e-test-utils';
+
+describe( 'Order of block keyboard navigation', () => {
+	beforeEach( async () => {
+		await createNewPost();
+	} );
+
+	it( 'permits tabbing through paragraph blocks in the expected order', async () => {
+		const paragraphBlocks = [ 'Paragraph 0', 'Paragraph 1', 'Paragraph 2' ];
+
+		const navigateToContentEditorTop = async () => {
+			// Use 'Ctrl+`' to return to the top of the editor
+			await page.keyboard.down( 'Control' );
+			await page.keyboard.press( '`' );
+			await page.keyboard.press( '`' );
+			await page.keyboard.up( 'Control' );
+			// Tab into the Title block
+			await page.keyboard.press( 'Tab' );
+		};
+
+		const tabThroughParagraphBlock = async ( paragraphText ) => {
+			// Tab to the next paragraph block
+			await page.keyboard.press( 'Tab' );
+
+			// The block external focusable wrapper has focus
+			const isFocusedParagraphBlock = await page.evaluate(
+				() => document.activeElement.dataset.type
+			);
+			await expect( isFocusedParagraphBlock ).toEqual( 'core/paragraph' );
+
+			// Tab causes 'add block' button to receive focus
+			await page.keyboard.press( 'Tab' );
+			const isFocusedParagraphInserterToggle = await page.evaluate( () =>
+				document.activeElement.classList.contains( 'editor-inserter__toggle' )
+			);
+			await expect( isFocusedParagraphInserterToggle ).toBe( true );
+
+			await tabThroughBlockMoverControl();
+			await tabThroughBlockToolbar();
+
+			// Tab causes the paragraph content to receive focus
+			await page.keyboard.press( 'Tab' );
+			const isFocusedParagraphContent = await page.evaluate(
+				() => document.activeElement.contentEditable
+			);
+			// The value of 'contentEditable' should be the string 'true'
+			await expect( isFocusedParagraphContent ).toBe( 'true' );
+
+			const paragraphEditableContent = await page.evaluate(
+				() => document.activeElement.innerHTML
+			);
+			await expect( paragraphEditableContent ).toBe( paragraphText );
+		};
+
+		const tabThroughBlockMoverControl = async () => {
+			// Tab to focus on the 'move up' control
+			await page.keyboard.press( 'Tab' );
+			const isFocusedMoveUpControl = await page.evaluate( () =>
+				document.activeElement.classList.contains( 'editor-block-mover__control' )
+			);
+			await expect( isFocusedMoveUpControl ).toBe( true );
+
+			// Tab to focus on the 'move down' control
+			await page.keyboard.press( 'Tab' );
+			const isFocusedMoveDownControl = await page.evaluate( () =>
+				document.activeElement.classList.contains( 'editor-block-mover__control' )
+			);
+			await expect( isFocusedMoveDownControl ).toBe( true );
+		};
+
+		const tabThroughBlockToolbar = async () => {
```

Comment:
Is there any way to automate this tabbing through the toolbar? It's going to be hard to maintain every time something changes in the paragraph block toolbar. It would be great to end up with a set of utilities which can be applied to every core block, not only paragraph.

Reply:
This is a really interesting idea, and I'd love to have a look at doing this over the next few days

Reply:
We can split this work into 2 PRs if you prefer. It's valuable as is if we add the code comment I mentioned in my previous note. Let me know how do you want to approach it.